pkg/certificate/transitions.go

```diff
@@ -0,0 +1,143 @@
+package certificate
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"k8s.io/client-go/util/retry"
+)
+
+type deferredCertificateStatusUpdate struct {
+	certificateTypes         []CertificateType
+	wait                     time.Duration
+	mrcClient                MRCClient
+	targetMRCName            string
+	targetCertificateStatus  v1alpha2.MeshRootCertificateComponentStatus
+	certificateStatusUpdates map[CertificateType]func() error
+}
+
+// a setFunc is a function that sets the certStatus of an MRC, returning true if the in-memory set was completed
+// and false if not. A return value of false indicates that the k8s update of the MRC resource should not occur
+type setFunc func(mrc *v1alpha2.MeshRootCertificate, status v1alpha2.MeshRootCertificateComponentStatus) bool
+
+func (m *Manager) toValidatingRollout(ctx context.Context, from string, mrc *v1alpha2.MeshRootCertificate, mrcClient MRCClient) error {
+	// Set the NewMRC's CA to the validating issuer
+	client, ca, err := mrcClient.GetCertIssuerForMRC(mrc)
+	if err != nil {
+		return err
+	}
+	c := &issuer{Issuer: client, ID: mrc.Name, CertificateAuthority: ca, TrustDomain: mrc.Spec.TrustDomain}
+	m.mu.Lock()
+	m.validatingIssuer = c
+	m.mu.Unlock()
+
+	// TODO: Set `wait` to real value from MRC once API changes are merged
+	// TODO: Change `targetCertificateStatus` to constant
+	d := newDeferredCertificateStatusUpdate(30*time.Minute, m.ownedCertTypes, mrcClient, mrc.Name, v1alpha2.MeshRootCertificateComponentStatus("Validating"))
+
+	go d.do(ctx)
+
+	return nil
+}
+
+func newDeferredCertificateStatusUpdate(wait time.Duration, certificateTypes []CertificateType, mrcClient MRCClient, targetMRCName string, targetCertStatus v1alpha2.MeshRootCertificateComponentStatus) *deferredCertificateStatusUpdate {
+	d := &deferredCertificateStatusUpdate{
+		wait:                    wait,
+		certificateTypes:        certificateTypes,
+		mrcClient:               mrcClient,
+		targetMRCName:           targetMRCName,
+		targetCertificateStatus: targetCertStatus,
+	}
+
+	d.certificateStatusUpdates = map[CertificateType]func() error{
```

Comment:
instead of this map, why not make these all receivers on a type, and have a switch case that calls each method by name based on the cert type? removes a layer of indirection, easier code nav, improved readability